metis-core/metis-core-service/src/main/java/eu/europeana/metis/core/execution/QueueConsumer.java

```diff
@@ -66,94 +70,123 @@ public QueueConsumer(Channel rabbitmqConsumerChannel, String rabbitmqQueueName,
+   */
   @Override
   public void handleDelivery(String consumerTag, Envelope rabbitmqEnvelope,
       AMQP.BasicProperties properties, byte[] body) throws IOException {
     String objectId = new String(body, StandardCharsets.UTF_8);
     LOGGER.info("WorkflowExecution id: {} received from queue.", objectId);
-    //Clean thread pool, some executions might have already finished
-    if (threadsCounter >= workflowExecutionSettings.getMaxConcurrentThreads()) {
-      LOGGER.debug(
-          "Trying to clean thread pool, found thread pool full with threadsCounter: {}, maxConcurrentThreads: {}",
-          threadsCounter, workflowExecutionSettings.getMaxConcurrentThreads());
-      checkAndCleanCompletionService();
-    }
 
-    boolean sendAck = true;
     try {
-      WorkflowExecution workflowExecution =
-          persistenceProvider.getWorkflowExecutionDao().getById(objectId);
+      WorkflowExecution workflowExecution = workflowExecutorManager.getWorkflowExecutionDao()
+          .getById(objectId);
       if (workflowExecution == null) {
         // This execution no longer exists and we need to ignore it.
         LOGGER.warn("Workflow execution with id: {} is in queue but no longer exists.", objectId);
       } else if (workflowExecution.isCancelling()) {
         // Has been cancelled, do not execute
         workflowExecution.setWorkflowAndAllQualifiedPluginsToCancelled();
-        persistenceProvider.getWorkflowExecutionDao().update(workflowExecution);
+        workflowExecutorManager.getWorkflowExecutionDao().update(workflowExecution);
         LOGGER.info("Cancelled inqueue user workflow execution with id: {}",
             workflowExecution.getId());
       } else {
-        sendAck = submitExecutionOrMarkNack(objectId);
+        submitExecution(objectId);
       }
     } catch (RuntimeException e) {
       LOGGER.error(
           "Exception occurred during submitting message from queue to a workflowExecution for id {}",
           objectId, e);
     } finally {
-      sendAckOrNack(rabbitmqEnvelope, objectId, sendAck);
+      sendAck(rabbitmqEnvelope, objectId);
     }
   }
 
-  private boolean submitExecutionOrMarkNack(String objectId) {
-    boolean sendAck;
-    //If the thread pool is still full, executions are still active. Flag to send the message back to the queue.
-    if (threadsCounter >= workflowExecutionSettings.getMaxConcurrentThreads()) {
-      sendAck = false;
-    } else {
-      // Submit for execution
-      WorkflowExecutor workflowExecutor = new WorkflowExecutor(objectId, persistenceProvider,
-          workflowExecutionSettings, workflowExecutionMonitor);
-      completionService.submit(workflowExecutor);
-      threadsCounter++;
-      sendAck = true;
-    }
-    return sendAck;
+  private void submitExecution(String objectId) {
+    WorkflowExecutor workflowExecutor = new WorkflowExecutor(objectId, workflowExecutorManager,
+        workflowExecutionSettings, workflowExecutionMonitor);
```

Comment:
Well the issue is that it has to go through the plugins and check if it can run per plugin and not just claim it like before. 
And to do so the semaphore should be acquired..
So even if one plugin is run the next one might not be allowed.
We can do what you say, but I thought it would complicate it more without much benefit?

Reply:
So after discussing this, I moved only the claiming functionality to the QueueConsumer